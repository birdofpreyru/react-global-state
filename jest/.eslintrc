{
  "env": {
    "browser": true,
    "jest": true
  },
  "extends": ["airbnb-typescript"],
  "parserOptions": {
    "project": "__tests__/tsconfig.json"
  },
  "plugins": ["jest"],
  "rules": {
    "global-require": 0,
    "import/no-extraneous-dependencies": 0,
    "no-console": 0
  }
}