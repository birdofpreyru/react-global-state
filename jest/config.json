{
  "collectCoverage": true,
  "collectCoverageFrom": ["src/**/*.{js,jsx,ts,tsx}"],
  "coverageDirectory": "__coverage__",
  "rootDir": "..",
  "testPathIgnorePatterns": [
    "/__assets__/",
    "/node_modules/"
  ],
  "transform": {
    "\\.(j|t)sx?$": "babel-jest"
  }
}
