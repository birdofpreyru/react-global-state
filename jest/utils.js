/**
 * Jest test utils.
 */

import _ from 'lodash';
import mockdate from 'mockdate';
import { render, unmountComponentAtNode } from 'react-dom';

export { act } from 'react-dom/test-utils';

/**
 * Generates a mock UUID.
 * @param {Number} seed
 * @return {String}
 */
export function getMockUuid(seed = 0) {
  const x = seed.toString(16).padStart(32, '0');
  return `${x.slice(0, 8)}-${x.slice(8, 12)}-${x.slice(12, 16)}-${x.slice(16, 20)}-${x.slice(20)}`;
}

const originalConsole = console.log;

export async function mockConsoleLog() {
  console.log = (...args) => console.log.logs.push(_.cloneDeep(args));
  console.log.clear = () => {
    console.log.logs = [];
  };
  console.log.clear();
}

export async function unMockConsoleLog() {
  console.log = originalConsole;
}

/**
 * Advances mock timers, and mock date by the specified time.
 * @param {Number} time Time step [ms].
 * @returns {Promise} Wait for this to "jump after" any async code which should
 *  be executed because of the mock time movement.
 */
export async function mockTimer(time) {
  mockdate.set(time + Date.now());
  jest.advanceTimersByTime(time);
}

/**
 * Mounts `Scene` to the DOM, and returns the root scene element.
 * @param {React.ReactNode} scene
 * @return {HTMLElement}
 */
export function mount(scene) {
  const res = document.createElement('div');
  document.body.appendChild(res);
  render(scene, res);
  return res;
}

/**
 * Waits for `time`.
 * @param {Number} time Time [ms].
 * @returns {Promise} Resolves once `time` have passed.
 */
export async function timer(time) {
  return new Promise((resolve) => {
    setTimeout(resolve, time);
  });
}

/**
 * Unmounts `scene` from the DOM.
 * @param {HTMLElement} scene
 */
export function unmount(scene) {
  unmountComponentAtNode(scene);
  scene.remove();
}
