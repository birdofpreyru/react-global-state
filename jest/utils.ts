/**
 * Jest test utils.
 */

import { cloneDeep } from 'lodash';
import mockdate from 'mockdate';
import { type ReactNode } from 'react';
import { type Root, createRoot } from 'react-dom/client';
import { act } from 'react-dom/test-utils';

export { act };

(global as { [key: string]: unknown }).IS_REACT_ACT_ENVIRONMENT = true;

/**
 * Generates a mock UUID.
 * @param seed
 * @return
 */
export function getMockUuid(seed = 0) {
  const x = seed.toString(16).padStart(32, '0');
  return `${x.slice(0, 8)}-${x.slice(8, 12)}-${x.slice(12, 16)}-${x.slice(16, 20)}-${x.slice(20)}`;
}

export const consoleLogs: unknown[] = [];

const originalConsole = console.log;

export async function mockConsoleLog() {
  console.log = (...args) => consoleLogs.push(cloneDeep(args));
  console.groupCollapsed = console.log;
  consoleLogs.splice(0);
}

export async function unMockConsoleLog() {
  console.log = originalConsole;
}

/**
 * Advances mock timers, and mock date by the specified time.
 * @param time Time step [ms].
 * @returns Wait for this to "jump after" any async code which should
 *  be executed because of the mock time movement.
 */
export async function mockTimer(time: number) {
  mockdate.set(time + Date.now());
  jest.advanceTimersByTime(time);
}

export type DestroyableHtmlElement = HTMLElement & { destroy: () => void };

/**
 * Mounts `Scene` to the DOM, and returns the root scene element.
 * @param scene
 * @return {HTMLElement} Created container DOM element with unmount
 *  function attached.
 */
export function mount(scene: ReactNode): DestroyableHtmlElement {
  let root: Root;

  const rootElement = document.createElement('div');

  const res: DestroyableHtmlElement = (rootElement as unknown) as DestroyableHtmlElement;

  res.destroy = () => {
    act(() => root.unmount());
    res.remove();
  };

  document.body.appendChild(rootElement);
  act(() => {
    root = createRoot(rootElement);
    root.render(scene);
  });
  return res;
}

/**
 * Advances mock timers by the specified time, and waits for any DOM updates
 * to complete.
 * @param time [miliseconds]
 * @return Wait for this.
 */
export function wait(time: number) {
  return act(() => mockTimer(time));
}
