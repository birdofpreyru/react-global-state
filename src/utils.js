// Auxiliary stuff.

/**
 * Returns 'true' if debug logging should be performed; 'false' otherwise.
 *
 * BEWARE: The actual safeguards for the debug logging still should read
 *  if (process.env.NODE_ENV !== 'production' && isDebugMode()) {
 *    // Some debug logging
 *  }
 * to ensure that debug code is stripped out by Webpack in production mode.
 *
 * @returns {boolean}
 * @ignore
 */
export function isDebugMode() {
  try {
    return process.env.NODE_ENV !== 'production'
      && !!process.env.REACT_GLOBAL_STATE_DEBUG;
  } catch (error) {
    return false;
  }
}

export default null;
