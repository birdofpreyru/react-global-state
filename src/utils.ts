import { type GetFieldType } from 'lodash';

export type CallbackT = () => void;

// TODO: This probably should be moved to JS Utils lib.
export type TypeLock<
  Unlocked extends 0 | 1,
  LockedT extends never | void,
  UnlockedT,
> = Unlocked extends 0 ? LockedT : UnlockedT;

/**
 * Given the type of state, `StateT`, and the type of state path, `PathT`,
 * it evaluates the type of value at that path of the state, if it can be
 * evaluated from the path type (it is possible when `PathT` is a string
 * literal type, and `StateT` elements along this path have appropriate
 * types); otherwise it falls back to the specified `UnknownT` type,
 * which should be set either `never` (for input arguments), or `void`
 * (for return types) - `never` and `void` in those places forbid assignments,
 * and are not auto-inferred to more permissible types.
 *
 * BEWARE: When StateT is any the construct resolves to any for any string
 * paths.
 */
export type ValueAtPathT<
  StateT,
  PathT extends null | string | undefined,
  UnknownT extends never | undefined | void,
> = unknown extends StateT
  ? UnknownT
  : string extends PathT
    ? UnknownT
    : PathT extends null | undefined
      ? StateT
      : GetFieldType<StateT, PathT> extends undefined
        ? UnknownT : GetFieldType<StateT, PathT>;

export type ValueOrInitializerT<ValueT> = ValueT | (() => ValueT);

/**
 * Returns 'true' if debug logging should be performed; 'false' otherwise.
 *
 * BEWARE: The actual safeguards for the debug logging still should read
 *  if (process.env.NODE_ENV !== 'production' && isDebugMode()) {
 *    // Some debug logging
 *  }
 * to ensure that debug code is stripped out by Webpack in production mode.
 *
 * @returns
 * @ignore
 */
export function isDebugMode(): boolean {
  try {
    return process.env.NODE_ENV !== 'production'
      && !!process.env.REACT_GLOBAL_STATE_DEBUG;
  } catch (error) {
    return false;
  }
}
