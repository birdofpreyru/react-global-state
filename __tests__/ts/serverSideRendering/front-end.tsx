/** @jest-environment jsdom */

// ^^^ although it is strange to do SSR test in JSDom environment,
// some of the tests should be done at the client side.

/**
 * Base test of the server-side rendering features.
 */

import mockdate from 'mockdate';
import pretty from 'pretty';
import { type JSXElementConstructor } from 'react';

import { type DestroyableHtmlElement } from 'jest/utils';

let JU: any;
let LIB: any;

jest.mock('uuid');

jest.useFakeTimers();
mockdate.set('2019-11-07Z');

let loaderA: jest.Mock;
let loaderB: jest.Mock;
let Scene: JSXElementConstructor<unknown>;
let scene: DestroyableHtmlElement | undefined;

beforeEach(() => {
  jest.resetModules();
  delete process.env.REACT_GLOBAL_STATE_DEBUG;
  JU = require('jest/utils');
  JU.unMockConsoleLog();
  ({ default: Scene, loaderA, loaderB } = require('./__assets__/TestScene'));
  loaderA.mockClear();
  loaderB.mockClear();
});

afterEach(() => {
  if (scene) {
    scene.destroy();
    scene = undefined;
  }

  // NOTE: Without this clearing some pending async updates from
  // "Scene test in the front-end mode" get fired only after the
  // later timer waiting in SSR test, which messes up the console
  // log snapshot verification.
  jest.clearAllTimers();
});

test('Scene test in the front-end mode', async () => {
  LIB = require('src');
  scene = JU.mount((
    <LIB.GlobalStateProvider>
      <Scene />
    </LIB.GlobalStateProvider>
  ));
  expect(pretty(scene!.innerHTML)).toMatchSnapshot();
  await JU.act(async () => {
    await JU.mockTimer(100);
  });
  expect(pretty(scene!.innerHTML)).toMatchSnapshot();
  expect(loaderA).toHaveBeenCalledTimes(1);
  expect(loaderB).toHaveBeenCalledTimes(0);
  await JU.act(async () => {
    await JU.mockTimer(1000);
  });
  await JU.act(async () => {
    await JU.mockTimer(0);
  });
  expect(pretty(scene!.innerHTML)).toMatchSnapshot();
  expect(loaderA).toHaveBeenCalledTimes(1);
  expect(loaderB).toHaveBeenCalledTimes(0);
});

describe('Test `getSsrContext()` function', () => {
  function SceneUsingSsrContext(
    { throwWithoutSsrContext }: { throwWithoutSsrContext?: boolean },
  ) {
    const ssrContext = LIB.getSsrContext(throwWithoutSsrContext);
    return (
      <div>
        {JSON.stringify(ssrContext, null, 2)}
      </div>
    );
  }

  SceneUsingSsrContext.defaultProps = {
    throwWithoutSsrContext: true,
  };

  let consoleError: typeof console.error;
  beforeAll(() => {
    consoleError = console.error;
  });

  afterEach(() => {
    console.error = consoleError;
    if (scene) {
      scene.destroy();
      scene = undefined;
    }
  });

  test('Missing GlobalStateProvider', () => {
    LIB = require('src');
    console.error = () => null;
    let message;
    try {
      JU.mount(<SceneUsingSsrContext />);
    } catch (error) {
      ({ message } = error as Error);
    }
    expect(message).toMatchSnapshot();
  });

  test('Get SSR context when exists', () => {
    LIB = require('src');
    scene = JU.mount((
      <LIB.GlobalStateProvider ssrContext={{ key: 'Dummy SSR Context' }}>
        <SceneUsingSsrContext />
      </LIB.GlobalStateProvider>
    ));
    expect(pretty(scene!.innerHTML)).toMatchSnapshot();
  });

  test('Get SSR context when does not exist', () => {
    LIB = require('src');
    console.error = () => null;
    let message;
    try {
      JU.mount((
        <LIB.GlobalStateProvider>
          <SceneUsingSsrContext />
        </LIB.GlobalStateProvider>
      ));
    } catch (error) {
      ({ message } = error as Error);
    }
    expect(message).toMatchSnapshot();
  });

  test('Get SSR context when does not exist, but no throw is opted in', () => {
    LIB = require('src');
    scene = JU.mount((
      <LIB.GlobalStateProvider>
        <SceneUsingSsrContext throwWithoutSsrContext={false} />
      </LIB.GlobalStateProvider>
    ));
    expect(pretty(scene!.innerHTML)).toMatchSnapshot();
  });
});
