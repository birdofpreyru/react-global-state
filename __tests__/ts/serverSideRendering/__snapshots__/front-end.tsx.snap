// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scene test in the front-end mode 1`] = `
"<div>
  <h1>defaultValue</h1>
  <div>{
    "data": null,
    "loading": true,
    "timestamp": 0
    }</div>
  <div>{
    "data": null,
    "loading": true,
    "timestamp": 0
    }</div>
</div>"
`;

exports[`Scene test in the front-end mode 2`] = `
"<div>
  <h1>defaultValue</h1>
  <div>{
    "data": null,
    "loading": true,
    "timestamp": 0
    }</div>
  <div>{
    "data": null,
    "loading": true,
    "timestamp": 0
    }</div>
</div>"
`;

exports[`Scene test in the front-end mode 3`] = `
"<div>
  <h1>defaultValue</h1>
  <div>{
    "data": "data",
    "loading": false,
    "timestamp": 1573084801100
    }</div>
  <div>{
    "data": "data",
    "loading": false,
    "timestamp": 1573084801100
    }</div>
</div>"
`;

exports[`Test \`getSsrContext()\` function Get SSR context when does not exist 1`] = `"No SSR context found"`;

exports[`Test \`getSsrContext()\` function Get SSR context when does not exist, but no throw is opted in 1`] = `"<div></div>"`;

exports[`Test \`getSsrContext()\` function Get SSR context when exists 1`] = `
"<div>{
  "key": "Dummy SSR Context",
  "dirty": false,
  "pending": []
  }</div>"
`;

exports[`Test \`getSsrContext()\` function Missing GlobalStateProvider 1`] = `"Missing GlobalStateProvider"`;
